skate-plugin/src/main/kotlin/com/slack/sgp/intellij/idemetrics/IndexingListener.kt

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (C) 2024 Slack Technologies, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.slack.sgp.intellij.idemetrics
+
+import com.intellij.util.indexing.diagnostic.ProjectIndexingHistory
+import com.intellij.util.indexing.diagnostic.ProjectIndexingHistoryListener
+import com.intellij.util.indexing.diagnostic.dto.toMillis
+import com.slack.sgp.intellij.tracing.IndexingEvent
+import com.slack.sgp.intellij.tracing.SkateSpanBuilder
+import com.slack.sgp.intellij.tracing.SkateTraceReporter
+import com.slack.sgp.intellij.util.isTracingEnabled
+
+@Suppress("UnstableApiUsage")
+class IndexingListener : ProjectIndexingHistoryListener {
+  override fun onFinishedIndexing(projectIndexingHistory: ProjectIndexingHistory) {
+    val project = projectIndexingHistory.project
+    if (!project.isTracingEnabled()) return
+    val skateSpanBuilder = SkateSpanBuilder()
+    skateSpanBuilder.apply {
+      projectIndexingHistory.indexingReason?.let {
+        addTag(IndexingEvent.INDEXING_REASON.name.take(200), it)
+      }
+      addTag(
+        IndexingEvent.UPDATING_TIME.name,
+        projectIndexingHistory.times.totalUpdatingTime.toMillis(),
+      )
+      addTag(
+        IndexingEvent.SCAN_FILES_DURATION.name,
+        projectIndexingHistory.times.scanFilesDuration.toMillis(),
+      )
+      addTag(
+        IndexingEvent.INDEXING_DURATION.name,
+        projectIndexingHistory.times.indexingDuration.toMillis(),
+      )
+      addTag(IndexingEvent.WAS_INTERRUPTED.name, projectIndexingHistory.times.wasInterrupted)
+      addTag(IndexingEvent.SCANNING_TYPE.name, projectIndexingHistory.times.scanningType.name)
+      addTag("event", IndexingEvent.INDEXING_COMPLETED.name)
+    }
+    SkateTraceReporter(project)
```

Comment:
Should we cache this reporter instance in a service somewhere?

Reply:
That's a really good point. I haven't really thought about that. Ended up trying to initialize the instance lazily for each class here https://github.com/slackhq/slack-gradle-plugin/pull/756/commits/14c05f8d2e153ee0b8e59c5469aaffbd02d8b635. Let me know if there's a better way to write this

Reply:
I think that's close, but we should put it somewhere more common than the whats new window I think? Does `SkateProjectService` work?

Reply:
Tried creating `SkateTraceService` here [here](https://github.com/slackhq/slack-gradle-plugin/blob/e8ebb38959cfe626382332675c7884f4669115ae/skate-plugin/src/main/kotlin/com/slack/sgp/intellij/tracing/SkateTraceService.kt). I ended up not using lazy initialization anymore in case if someone opened another repo from the IDE, then we need to reinitialize the trace reporter